### DISCUSSION TITLE
从源代码看 ObjC 中消息的发送 · /message

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#294] [Announcements]
https://draveness.me/message

**Comment by Robinguo2012** [comment]
博主是自己编译了runtime 库吗? 如何在objc-runtime-new.mm 中打断点呢?

**Comment by draveness** [comment]
> @Robinguo2012
> 博主是自己编译了runtime 库吗? 如何在objc-runtime-new.mm 中打断点呢?

你可以看下这个，不知道现在还能用么 https://github.com/RetVal/objc-runtime

**Comment by Robinguo2012** [comment]
可以用,谢楼主!

**Comment by Robinguo2012** [comment]
博主你好,有两个问题请教
1. 在打印buckets时,如何知道objc_class 数据成员 cache_t 的地址?
   (objc_class *) $0 = 0x0000000100001230
   (lldb) p (cache_t *)0x0000000100001240
   (cache_t *) $1 = 0x0000000100001240

2. method_array_t 好像不是一个二维数组,而是一个继承自模板类list_array_tt 的C++ 类;

**Comment by draveness** [comment]
> @Robinguo2012
> 博主你好,有两个问题请教
> 1. 在打印buckets时,如何知道objc_class 数据成员 cache_t 的地址?
>    (objc_class *) $0 = 0x0000000100001230
>    (lldb) p (cache_t *)0x0000000100001240
>    (cache_t *) $1 = 0x0000000100001240
> 
> 2. method_array_t 好像不是一个二维数组,而是一个继承自模板类list_array_tt 的C++ 类;

第一个问题我也忘了，第二个问题你可以用 C++ API 吧？

**Comment by xuzhongping** [comment]
> @Robinguo2012
> 博主你好,有两个问题请教
> 1. 在打印buckets时,如何知道objc_class 数据成员 cache_t 的地址?
>    (objc_class *) $0 = 0x0000000100001230
>    (lldb) p (cache_t *)0x0000000100001240
>    (cache_t *) $1 = 0x0000000100001240
> 
> 2. method_array_t 好像不是一个二维数组,而是一个继承自模板类list_array_tt 的C++ 类;

我在测试中也是发现objc_class *的地址偏移0x10，可以得到成员cache的地址，并且使用后能获取到需要的信息

**Comment by kobezou** [comment]
博主，想问下，你那个代码是怎么调试的呀？我这边下载后可以运行起来，但是debug不走里面的代码？是需要手动调用一下吗？

**Comment by ruoxuecf** [comment]
> @Robinguo2012
> 博主你好,有两个问题请教
> 1. 在打印buckets时,如何知道objc_class 数据成员 cache_t 的地址?
>    (objc_class *) $0 = 0x0000000100001230
>    (lldb) p (cache_t *)0x0000000100001240
>    (cache_t *) $1 = 0x0000000100001240
> 
> 2. method_array_t 好像不是一个二维数组,而是一个继承自模板类list_array_tt 的C++ 类;

这个看objc_class的结构体组成就知道了
struct objc_class : objc_object {
    isa_t isa;
    Class superclass;
    cache_t cache;
    class_data_bits_t bits;
};
这个里面isa和superClass都是8个字节, 加在一起是16个字节, objc_class地址 + 0x10就是cache的地址

**Comment by ruoxuecf** [comment]
看到这个学会了很多调试的技巧, 收益匪浅, 感谢博主.
最后想问下视频的链接地址在哪?
